zerver/lib/streams.py

```diff
@@ -179,6 +180,18 @@ def get_default_values_for_stream_permission_group_settings(
     return group_setting_values
 
 
+def get_user_ids_with_metadata_access_via_permission_groups(stream: Stream) -> set[int]:
+    stream_admin_user_ids = set(
+        get_recursive_group_members(stream.can_administer_channel_group).values_list(
+            "id", flat=True
+        )
+    )
+    stream_add_subscribers_group_user_ids = set(
+        get_recursive_group_members(stream.can_add_subscribers_group).values_list("id", flat=True)
+    )
+    return stream_admin_user_ids | stream_add_subscribers_group_user_ids
```

Comment:
This can likely be further optimized by having a variant of `get_recursive_group_members` that accepts multiple groups as its input.

Reply:
Optimized in c6e3410f48de51f891d74c9d359072658bf79690. The PR is a bit of a mess in the sense that some optimisations are their own commits, but otherwise the commits were becoming too big for me to make sense of.

Reply:
It's fine to optimize in later commits!